File: src/test/java/com/soapdogg/lispInterpreter/valueretriver/AtomicValueRetrieverTest.java
package com.soapdogg.lispInterpreter.valueretriver;

import com.soapdogg.lispInterpreter.constants.FunctionNameConstants;
import com.soapdogg.lispInterpreter.determiner.ExpressionNodeDeterminer;
import com.soapdogg.lispInterpreter.exceptions.NotAtomicException;
import com.soapdogg.lispInterpreter.printer.ListNotationPrinter;
import com.soapdogg.lispInterpreter.datamodels.AtomNode;
import com.soapdogg.lispInterpreter.datamodels.Node;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

class AtomicValueRetrieverTest {

    private Node node;
    private int position;
    private String functionName;

    private ExpressionNodeDeterminer expressionNodeDeterminer;
    private AtomicValueRetriever atomicValueRetriever;

    @BeforeEach
    void setup() {
        node = Mockito.mock(AtomNode.class);
        position = 1;
        functionName = FunctionNameConstants.TIMES;

        expressionNodeDeterminer = Mockito.mock(ExpressionNodeDeterminer.class);
        ListNotationPrinter listNotationPrinter = Mockito.mock(ListNotationPrinter.class);
        atomicValueRetriever = AtomicValueRetriever.newInstance(
            expressionNodeDeterminer,
            listNotationPrinter
        );
    }

    @Test
    void nodeIsListTest() {
        Mockito.when(expressionNodeDeterminer.isExpressionNode(node)).thenReturn(true);

        Assertions.assertThrows(
            NotAtomicException.class,
            () -> atomicValueRetriever.retrieveAtomicValue(
                node,
                position,
                functionName
            )
        );
    }

    @Test
    void nodeIsNotListTest() throws Exception {
        Mockito.when(expressionNodeDeterminer.isExpressionNode(node)).thenReturn(false);

        String value = "value";
        Mockito.when(((AtomNode)node).getValue()).thenReturn(value);

        String actual = atomicValueRetriever.retrieveAtomicValue(
            node,
            position,
            functionName
        );

        Assertions.assertEquals(value, actual);
    }
}

File: src/test/java/com/soapdogg/lispInterpreter/valueretriver/AtomicValueRetrieverTest.kt
package com.soapdogg.lispInterpreter.valueretriver

import com.soapdogg.lispInterpreter.constants.FunctionNameConstants
import com.soapdogg.lispInterpreter.datamodels.AtomNode
import com.soapdogg.lispInterpreter.datamodels.ExpressionNode
import com.soapdogg.lispInterpreter.exceptions.NotAtomicException
import com.soapdogg.lispInterpreter.printer.ListNotationPrinter
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.Mockito

class AtomicValueRetrieverTest {
    private val position = 1
    private val functionName = FunctionNameConstants.TIMES
    private val listNotationPrinter = Mockito.mock(ListNotationPrinter::class.java)
    private val atomicValueRetriever = AtomicValueRetriever(
        listNotationPrinter
    )

    @Test
    fun nodeIsListTest() {
        val node = Mockito.mock(ExpressionNode::class.java)
        Assertions.assertThrows(
            NotAtomicException::class.java
        ) {
            atomicValueRetriever.retrieveAtomicValue(
                node,
                position,
                functionName
            )
        }
    }

    @Test
    fun nodeIsNotListTest() {
        val node = Mockito.mock(AtomNode::class.java)
        val value = "value"
        Mockito.`when`(node.value).thenReturn(value)
        val actual = atomicValueRetriever.retrieveAtomicValue(
            node,
            position,
            functionName
        )
        Assertions.assertEquals(value, actual)
    }
}